android/src/toga_android/fonts.py

```diff
@@ -28,69 +28,69 @@ class Font:
-
-        if typeface is None:
-            if self.interface.family is SYSTEM:
-                # The default button font is not marked as bold, but it has a weight
-                # of "medium" (500), which is in between "normal" (400), and "bold"
-                # (600 or 700). To preserve this, we use the widget's original
-                # typeface as a starting point rather than Typeface.DEFAULT.
-                typeface = default
-            elif self.interface.family is MESSAGE:
-                typeface = Typeface.DEFAULT
-            elif self.interface.family is SERIF:
-                typeface = Typeface.SERIF
-            elif self.interface.family is SANS_SERIF:
-                typeface = Typeface.SANS_SERIF
-            elif self.interface.family is MONOSPACE:
-                typeface = Typeface.MONOSPACE
-            elif self.interface.family is CURSIVE:
-                typeface = Typeface.create("cursive", Typeface.NORMAL)
-            elif self.interface.family is FANTASY:
-                # Android appears to not have a fantasy font available by default,
-                # but if it ever does, we'll start using it. Android seems to choose
-                # a serif font when asked for a fantasy font.
-                typeface = Typeface.create("fantasy", Typeface.NORMAL)
-            else:
-                typeface = Typeface.create(self.interface.family, Typeface.NORMAL)
-
-        native_style = typeface.getStyle()
+                try:
+                    font_path = _REGISTERED_FONT_CACHE[font_key]
+                except KeyError:
+
+                    # 3a. Not a user-registered font
+                    raise UnknownFontError(f"Unknown font '{self.interface}'")
+
+                # 3b. User has indeed registered this font.
+                else:
+                    if Path(font_path).is_file():
+                        self.native_typeface = Typeface.createFromFile(font_path)
+                        if self.native_typeface is Typeface.DEFAULT:
+                            raise ValueError(f"Unable to load font file {font_path}")
+                    else:
+                        raise ValueError(f"Font file {font_path} could not be found")
+
+            _FONT_CACHE[self.interface] = self.native_typeface
+
+        self.native_style = 0
         if self.interface.weight == BOLD:
-            native_style |= Typeface.BOLD
+            self.native_style |= Typeface.BOLD
         if self.interface.style in {ITALIC, OBLIQUE}:
-            native_style |= Typeface.ITALIC
+            self.native_style |= Typeface.ITALIC
+
+    def typeface(self, default=Typeface.DEFAULT):
+        typeface = default if self.native_typeface == SYSTEM else self.native_typeface
 
-        if native_style != typeface.getStyle():
-            typeface = Typeface.create(typeface, native_style)
+        if self.native_style != typeface.getStyle():
+            typeface = Typeface.create(typeface, self.native_style)
```

Comment:
At first I was concerned that the return value of `Typeface.create` is never added to `_FONT_CACHE`. But then I looked at the [Typeface source code](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/graphics/java/android/graphics/Typeface.java), and found that Android already has its own cache for different styles of the same Typeface. It would be worth adding a comment saying this.